File: src/lib/scala/Slate.Core/src/main/scala/slate/core/common/EntityServiceWithSupport.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.core.common

import slate.common.encrypt.EncryptSupportIn
import slate.common.i18n.I18nSupportIn
import slate.common.logging.LogSupportIn
import slate.common.results.ResultSupportIn
import slate.entities.core.{EntityRepo, EntityService, Entity}

/**
 * Entity Service wrapper with support for encryption, logging, results, and application context
 * @param context
 * @param repo
 * @tparam T
 */
class EntityServiceWithSupport[T >: Null <: Entity](val context:AppContext, repo:EntityRepo[T])
  extends EntityService[T](repo)
  with EncryptSupportIn
  with LogSupportIn
  with I18nSupportIn
  with ResultSupportIn
  with AppContextSupport
{

  def initContext():Unit = {
  }


  override protected def log() = Option(context.log)
  override protected def enc() = context.enc
  override protected def res() = context.res

}

File: src/lib/kotlin/slatekit-entities/src/main/kotlin/slatekit/entities/support/EntityServiceWithSupport.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.entities.support

import slatekit.common.Context
import slatekit.common.encrypt.EncryptSupport
import slatekit.common.encrypt.Encryptor
import slatekit.common.log.LogSupport
import slatekit.common.log.Logger
import slatekit.entities.core.Entities
import slatekit.entities.core.Entity
import slatekit.entities.core.EntityRepo
import slatekit.entities.core.EntityService
import slatekit.results.Notice

/**
 * Entity Service wrapper with support for encryption, logging, results, and application context
 * @param context
 * @param repo
 * @tparam T
 */
open class EntityServiceWithSupport<T>(val context: Context, val entities: Entities, repo: EntityRepo<T>)
    : EntityService<T>(entities, repo), EncryptSupport, LogSupport where T : Entity {

    override val logger: Logger? get() = context.logs.getLogger()
    override val encryptor: Encryptor? get() = context.enc

    protected fun <T> handleError(err: String): Notice<T> {
        logger?.error(err)
        return slatekit.results.Failure(err)
    }
}
